docs/component-specs.md

```diff
@@ -27,3 +80,103 @@ The variables that can be accessed by a condition are:
 - `runtime.family`: OS family, e.g. `"debian"`, `"redhat"`, `"windows"`, `"darwin"`
 - `runtime.major`, `runtime.minor`: the operating system version. Note that these are strings not integers, so they must be converted in order to use numeric comparison. For example to check if the OS major version is at most 12, use `number(runtime.major) <= 12`.
 - `user.root`: true if Agent is being run with root / administrator permissions.
+
+### `command` (required for shipper)
+
+The `command` field determines how the component will be run. Shippers must include this field, while inputs must include either `command` or `service`. `command` consists of the following subfields:
+
+#### `command.args` (list of strings)
+
+the command-line arguments to pass to this component when running it.
+
+#### `command.env`
+
+A list of environment variables to set when running this component. Each entry of the list consists of `name` and `value` pairs, for example:
+
+```yml
+command:
+  ...
+  env:
+    - name: DEBUG
+      value: "true"
+    - name: ELASTICSEARCH_HOST
+      value: "https://127.0.0.1:9200"
+```
+
+#### `command.timeouts`
+
+The timeout duration for various actions performed on this component:
+
+- `checkin`: Agent checkins
+- `restart`: Restarting the component
+- `stop`: Stopping the component
+
+For example:
+
+```yml
+command:
+  ...
+  timeouts:
+    checkin: 10s
+    restart: 30s
+```
+
+#### `command.log`
+
+Agent expects commands it runs to write their logs to standard output as lines of JSON. Each log event has certain standard data like log level and timestamp, however the keys for these values may vary between different programs. `command.log` specifies the meaning of JSON log events. It has the following subfields:
+
+- `level_key`: the JSON key for the event's log level
+- `time_key`: the JSON key for the event's timestamp
+- `time_format`: The format used for log timestamps. This field uses the definitions from [Golang's time formatting support](https://pkg.go.dev/time#Time.Format), which consists of providing an example string in the target format. Example values: `Mon, 02 Jan 2006 15:04:05 -0700`, `Mon Jan _2 15:04:05 MST 2006`.
+- `message_key`: the JSON key for the log message
+- `ignore_keys`: a list of JSON keys that should be skipped when reading log events from this command
+
+#### `restart_monitoring_period` (duration), `maximum_restarts_per_period` (integer)
+
+When a process fails, Agent will restart it. These parameters provide a rate limit, so Agent will not try to start the same component too many times in a short duration. If Agent restarts a process `maximum_restarts_per_period` times within `restart_monitoring_period`, then it will not try again until the monitoring period elapses.
```

Comment:
We actually use this to avoid marking Beats that restart one time in a period as failed in the UI, because they restart themselves each time the output configuration changes. See https://github.com/elastic/elastic-agent/pull/2003 for some context.

Reply:
Ah, I misunderstood the purpose of the calling function, that makes more sense now. Rewrote this section.